src/main/java/com/google/devtools/build/lib/query2/query/GraphlessBlazeQueryEnvironment.java

```diff
@@ -416,36 +422,55 @@ private Target getTargetOrThrow(Label label)
       throws QueryException {
     ThreadSafeMutableSet<Target> dependentFiles = createThreadSafeMutableSet();
     Set<PackageIdentifier> seenPackages = new HashSet<>();
+    Set<String> seenRepos = new HashSet<>();
 
     // Adds all the package definition files (BUILD files and build extensions) for package "pkg",
     // to dependentFiles.
     for (Target x : nodes) {
+      List<Label> extensions = new ArrayList<>();
       Package pkg = x.getPackage();
-      if (seenPackages.add(pkg.getPackageIdentifier())) {
+      if (x instanceof Rule && pkg.getPackageIdentifier().getCanonicalForm().equals("//external")) {
```

Comment:
This will still be incomplete if you have a non-Rule target in a Package, right? You can get a situation with `exports_files([...])` or just if the query restricts the output to source files or output files. The package will still be loaded and require those bzl files to work correctly (and since `buildfiles` is often used to check if a package needs to be reloaded, that's important), but those bzl files won't be listed here. That makes me think you're doing this at a fundamentally wrong place, although I don't have a better one to suggest since I'm not familiar with external repos. In any case, please add a test for these situations: querying for a source file exported via `exports_files([...])`; querying for an output file (the output of a genrule, not the genrule itself); and a query that filters for source files (`kind('source_file',...)`).

Reply:
This is a special case for `//external`. The only non-rule target there is `WORKSPACE`. For example, from Bazel's own repo:

```
bazel query 'kind("source file", //external:all-targets)'
//external:WORKSPACE
```

For `genrule` and `exports_files` in external repos, their labels are like `@some_repo//:data.txt`, which is not in the `//external` package, and out of scope of this pull request. I can write tests to cover them, but their behavior in `buildfiles` is not affected by this pull request.

Reply:
Oh, I see. Thanks, you can disregard this comment in that case.